### DISCUSSION TITLE
Software linked against libgccjit no longer works after upgrade to 14.1.0

### DISCUSSION THREAD

**Discussion by worr** [Homebrew/discussions#5364] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.21-117-gd7e81be
ORIGIN: https://github.com/Homebrew/brew
HEAD: d7e81bed359e1cbe9fb935ff40e1fd85269388fb
Last commit: 16 hours ago
Core tap HEAD: 55a464718370a37a8778a30531bbd5c35d1ea07e
Core tap last commit: 5 months ago
Core tap JSON: 13 May 17:57 UTC
Core cask tap HEAD: d1af27d453ab30275a215c9e5f84f9962cf2a91b
Core cask tap last commit: 5 months ago
Core cask tap JSON: 13 May 17:58 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BAT_THEME: Monokai Extended Light
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: emacsclient
HOMEBREW_MAKE_JOBS: 12
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.45.0 => /usr/local/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4.1-x86_64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  pygments
  python-tabulate

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  fish_add_path /usr/local/sbin
```

### Description of issue

After the update to 14.1.0, when I try and use anything linked against libgccjit, I get crashes trying to link against libgcc_s.dylib.

Here's a snippet of an Apple report from trying to run emacs-mac built to use native compilation (which uses libgccjit):

```
Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: @rpath/libgcc_s.1.1.dylib
Referenced from: <C93A88B4-A5C5-3C6F-9F9B-CB9D58FF003E> /usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.0.dylib
Reason: tried: '/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgcc_s.1.1.dylib' (no such file), '/usr/local/lib/libgcc_s.1.1.dylib' (no such file), '/usr/lib/libgcc_s.1.1.dylib' (no such file, not in dyld cache)
(terminated at launch; ignore backtrace)
```

If we look at the output of `otool -L`, we can see that it has an rpath entry for `libgcc_s.1.1.dylib` 

```
~|15:50:33|0$ otool -L /usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.0.dylib
/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.0.dylib:
	/usr/local/opt/libgccjit/lib/gcc/current/libgccjit.0.dylib (compatibility version 0.0.0, current version 26.0.26)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/local/opt/isl/lib/libisl.23.dylib (compatibility version 27.0.0, current version 27.0.0)
	/usr/local/opt/libmpc/lib/libmpc.3.dylib (compatibility version 7.0.0, current version 7.1.0)
	/usr/local/opt/mpfr/lib/libmpfr.6.dylib (compatibility version 9.0.0, current version 9.1.0)
	/usr/local/opt/gmp/lib/libgmp.10.dylib (compatibility version 16.0.0, current version 16.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	/usr/local/opt/zstd/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.5.6)
	@rpath/libgcc_s.1.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.100.2)
```

libgccjit doesn't include this

```
/usr/local/Cellar/libgccjit/14.1.0/INSTALL_RECEIPT.json
/usr/local/Cellar/libgccjit/14.1.0/.brew/libgccjit.rb
/usr/local/Cellar/libgccjit/14.1.0/ChangeLog
/usr/local/Cellar/libgccjit/14.1.0/include/libgccjit++.h
/usr/local/Cellar/libgccjit/14.1.0/include/libgccjit.h
/usr/local/Cellar/libgccjit/14.1.0/sbom.spdx.json
/usr/local/Cellar/libgccjit/14.1.0/README
/usr/local/Cellar/libgccjit/14.1.0/COPYING.LIB
/usr/local/Cellar/libgccjit/14.1.0/COPYING
/usr/local/Cellar/libgccjit/14.1.0/COPYING.RUNTIME
/usr/local/Cellar/libgccjit/14.1.0/NEWS
/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.dylib
/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.0.dylib
/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/14/libgccjit.dylib
/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/14/libgccjit.0.dylib
```

**Comment by Bo98** [comment]
Hmm yeah seems like libgccjit 13 didn't have the libgcc_s dependency but libgccjit 14 does. Maybe from https://github.com/gcc-mirror/gcc/blob/releases/gcc-14.1.0/gcc/jit/Make-lang.in#L161-L162

cc @fxcoudert

**Reply by fxcoudert** [reply]
`libgcc_s` is provided by `gcc`, do this get fixed if you `brew install gcc`?

**Reply by carlocab** [reply]
Probably won't be enough, since `libgccjit.0.dylib` will need an `LC_RPATH` pointing to `$HOMEBREW_PREFIX/opt/gcc/lib/gcc/current`

**Reply by fxcoudert** [reply]
I think with gcc it does work, since the libgccjit test passes. I tried it locally and gcc itself inserts the right entry.

**Reply by Bo98** [reply]
Seems like it's causing some confusion elsewhere too: https://stackoverflow.com/questions/78455650/library-libgccjit-not-loaded

Is `-lgccjit` supported with Clang (prior to GCC 14)? I suspect some may have been using that.

**Reply by carlocab** [reply]
I think it's used by this tap: https://github.com/d12frosted/homebrew-emacs-plus/blob/61a1e971b228f416a95374cb8e5d95631fff3a42/Formula/emacs-plus%4030.rb#L62

They have a `gcc => :build` dependency in the formula, but I don't see where they force the use of GCC to compile. Does that happen with `env :std`?

**Reply by Bo98** [reply]
> Does that happen with `env :std`?

All bets are off with `env :std`. It might, or it might not. Totally depends on what the upstream build scripts call - but it will need to be `gcc-14` if so and not just `gcc` which will still invoke Apple Clang. Given the recent bug reports, there's a good chance it's actually using Clang - and it may have worked before with libgccjit 13.

**Reply by worr** [reply]
I think @carlocab is right here. I have had `gcc@14` installed on this computer this whole time. Specifically, the app that's failing for me is `emacs-mac`, and if we look at the crash report:

```
Library not loaded: @rpath/libgcc_s.1.1.dylib
Referenced from: <C93A88B4-A5C5-3C6F-9F9B-CB9D58FF003E> /usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgccjit.0.dylib
Reason: tried: '/usr/local/Cellar/libgccjit/14.1.0/lib/gcc/current/libgcc_s.1.1.dylib' (no such file), '/usr/local/lib/libgcc_s.1.1.dylib' (no such file), '/usr/lib/libgcc_s.1.1.dylib' (no such file, not in dyld cache)
(terminated at launch; ignore backtrace)
```

we can see that it's not searching a path that'd ever have `libgcc_s` in it.

**Reply by carlocab** [reply]
Here's a possible workaround:
```
brew install gcc
install_name_tool -add_rpath "$(brew --prefix gcc)/lib/gcc/current" "$(brew --prefix libgccjit)/lib/gcc/current/libgccjit.0.dylib"
```

**Reply by fxcoudert** [reply]
So now libgccjit needs libgcc_s, but we can't ship libgcc_s with libgccjit because that would make it conflict with the gcc formula. One option (I don't like it) is to merge back the two formulas into a full-featured GCC.

But that is heavier to download for people who just use libgccjit, and slower to use for people who don't want JIT.

**Reply by worr** [reply]
The workaround works perfectly, definitely seems to point to that it needs an additional `LC_RPATH` entry. Thanks a ton

**Reply by edmwagner** [reply]
This works! Thank you @carlocab .


I was hitting the same issue with my emacs-29 build from source.

**Reply by carlocab** [reply]
> slower to use for people who don't want JIT.

How much slower? And even then, it's really only slower on Linux, since `--enable-host-shared` is enabled by default on macOS.

**Reply by fxcoudert** [reply]
Upstream recommendation: https://gcc.gnu.org/onlinedocs/gcc-14.1.0/jit/internals/index.html

> **Packaging notes**
> The configure-time option [--enable-host-shared](https://gcc.gnu.org/onlinedocs/gcc-14.1.0/jit/internals/index.html#cmdoption-enable-host-shared) is needed when building the jit in order to get position-independent code. This will slow down the regular compiler by a few percent. Hence when packaging gcc with libgccjit, please configure and build twice:
> - once without [--enable-host-shared](https://gcc.gnu.org/onlinedocs/gcc-14.1.0/jit/internals/index.html#cmdoption-enable-host-shared) for most languages, and
> - once with [--enable-host-shared](https://gcc.gnu.org/onlinedocs/gcc-14.1.0/jit/internals/index.html#cmdoption-enable-host-shared) for the jit

Yes, it's only on Linux, and only “a few percent”. On the other hand, GCC bootstrap is already long, and we probably don't want to build twice 😢 

Maybe we make `gcc` a dependency and hack the `libgccjit` rpath. Or we merge back the two, and incur the hit in runtime performance on Linux. I don't have a strong opinion.

**Reply by carlocab** [reply]
> Maybe we make `gcc` a dependency and hack the `libgccjit` rpath.

Yea, been contemplating that. Might be the "nicest", since it seems to only be necessary on Intel macOS. Strangely, `libgccjit` does not link to `libgcc_s` on M1 macOS (at least, not on Sonoma).

**Reply by fxcoudert** [reply]
Tried something in https://github.com/Homebrew/homebrew-core/pull/171822

**Comment by carlocab** [comment, marked as answer]
This should be fixed now. Do `brew update && brew reinstall libgccjit` in about ~20 minutes.

**Reply by worr** [reply]
Sorry, was away from work for awhile, and now can confirm that this works on my work laptop. Thanks so much for the quick resolution.
